packages/flutter/lib/src/rendering/wrap.dart

```diff
@@ -663,22 +665,114 @@ class RenderWrap extends RenderBox
+    final double mainAxisLimit;
+    final double Function(RenderBox child, double crossAxisSize) getChildMinIntrinsicMainAxisExtent;
+    final BoxConstraints Function(double maxSize) childConstraintsFittingLooseInRun, childConstraintsFittingTightInRun;
+
+    switch (direction) {
+      case Axis.horizontal:
+        childConstraintsLoose = BoxConstraints(maxWidth: constraints.maxWidth);
+        childConstraintsTight = BoxConstraints(minWidth: constraints.maxWidth, maxWidth: constraints.maxWidth);
+        mainAxisLimit = constraints.maxWidth;
+        getChildMinIntrinsicMainAxisExtent = (RenderBox child, double crossAxisSize) => child.getMinIntrinsicWidth(crossAxisSize);
+        childConstraintsFittingLooseInRun = (double maxSize) => BoxConstraints(maxWidth: maxSize);
+        childConstraintsFittingTightInRun = (double maxSize) => BoxConstraints(minWidth: maxSize, maxWidth: maxSize);
+      case Axis.vertical:
+        childConstraintsLoose = BoxConstraints(maxHeight: constraints.maxHeight);
+        childConstraintsTight = BoxConstraints(minHeight: constraints.maxHeight, maxHeight: constraints.maxHeight);
+        mainAxisLimit = constraints.maxHeight;
+        getChildMinIntrinsicMainAxisExtent = (RenderBox child, double crossAxisSize) => child.getMinIntrinsicHeight(crossAxisSize);
+        childConstraintsFittingLooseInRun = (double maxSize) => BoxConstraints(maxHeight: maxSize);
+        childConstraintsFittingTightInRun = (double maxSize) => BoxConstraints(minHeight: maxSize, maxHeight: maxSize);
+    }
 
     final (bool flipMainAxis, _) = _areAxesFlipped;
     final double spacing = this.spacing;
     final List<_RunMetrics> runMetrics = <_RunMetrics>[];
 
+    ({RenderBox current, RenderBox? next})? childWithNext(RenderBox? child) => child != null ? (current: child, next: childAfter(child)) : null;
+
     _RunMetrics? currentRun;
     _AxisSize childrenAxisSize = _AxisSize.empty;
-    for (RenderBox? child = firstChild; child != null; child = childAfter(child)) {
-      final _AxisSize childSize = _AxisSize.fromSize(size: layoutChild(child, childConstraints), direction: direction);
+    /// This caches the minIntrinsicMainAxisExtent of the next child, in case of WrapFit.maybeRunTight.
+    double? nextChildMinIntrinsicMainAxisExtent;
+    for (({RenderBox current, RenderBox? next})? child = childWithNext(firstChild); child != null; child = childWithNext(child.next)) {
+      // This lazily evaluates the minIntrinsicMainAxisExtent of the current child and uses the cache, if there is one.
+      final double? $cachedChildMinIntrinsicMainAxisExtent =
+          nextChildMinIntrinsicMainAxisExtent;
+      nextChildMinIntrinsicMainAxisExtent = null;
+      final ({RenderBox current, RenderBox? next}) $child = child;
+      late final double childMinIntrinsicMainAxisExtent =
+          $cachedChildMinIntrinsicMainAxisExtent ??
+              getChildMinIntrinsicMainAxisExtent($child.current, double.infinity);
```

Comment:
A few things here:
- This section might be easier to read with less verbose variable names, e.g. `nextChildMinExtent` instead of `nextChildMinIntrinsicMainAxisExtent`.
- It looks to me like we can just use `child` directly instead of creating a `$child` variable (but let me know if I'm wrong here).
- Using a `$` for a variable name isn't really conventional within the Flutter framework. Could you switch to names that don't have a `$`?

Reply:
Better? (3611c0a3f7)

I used the $child, to ensure nullsafety in the late evaluation. (I've added a comment explaining this pattern.)